File: src/stores/sandbox.ts
import { create } from "zustand"
import { invoke } from "@tauri-apps/api/core"
import { listen, type UnlistenFn } from "@tauri-apps/api/event"
import type { SandboxStatus, DownloadProgress, AgentInfo } from "@/types"

interface SandboxState {
  status: SandboxStatus
  progress: DownloadProgress | null
  bootLog: string[]
  error: string | null
  agentInfo: AgentInfo | null

  init: () => Promise<UnlistenFn[]>
  start: () => Promise<void>
  stop: () => Promise<void>
  restart: () => Promise<void>
  setEnabled: (enabled: boolean) => Promise<void>
  mountWorkspace: (path: string) => Promise<void>
  cleanWorkspace: () => Promise<void>
}

export const useSandboxStore = create<SandboxState>((set) => ({
  status: "not_installed",
  progress: null,
  bootLog: [],
  error: null,
  agentInfo: null,

  init: async () => {
    try {
      const state = await invoke<{ status: string; [key: string]: unknown }>("sandbox_get_state")
      set({ status: state.status as SandboxStatus })
    } catch {
      // Backend may not have registered sandbox commands yet
    }

    const unlisteners: UnlistenFn[] = []

    unlisteners.push(
      await listen<{ status: SandboxStatus; message?: string }>("sandbox:state-changed", (event) => {
        const { status, message } = event.payload
        set({
          status,
          error: status === "error" ? (message ?? "Unknown error") : null,
        })
      })
    )

    unlisteners.push(
      await listen<DownloadProgress>("sandbox:download-progress", (event) => {
        set({ progress: event.payload })
      })
    )

    unlisteners.push(
      await listen<string>("sandbox:boot-log", (event) => {
        set((state) => ({
          bootLog: [...state.bootLog, event.payload],
        }))
      })
    )

    unlisteners.push(
      await listen<AgentInfo>("sandbox:agent-ready", (event) => {
        set({ agentInfo: event.payload })
      })
    )

    return unlisteners
  },

  start: async () => {
    await invoke("sandbox_start")
  },

  stop: async () => {
    await invoke("sandbox_stop")
  },

  restart: async () => {
    set({ bootLog: [], error: null })
    await invoke("sandbox_restart")
  },

  setEnabled: async (enabled: boolean) => {
    await invoke("sandbox_set_enabled", { enabled })
  },

  mountWorkspace: async (path: string) => {
    await invoke("sandbox_mount_workspace", { path })
  },

  cleanWorkspace: async () => {
    await invoke("sandbox_clean_workspace")
  },
}))

File: src-tauri/src/commands/sandbox.rs
use tauri::{AppHandle, State};

use crate::sandbox::manager::SandboxManager;
use crate::sandbox::{ExecResult, SandboxState};

#[tauri::command]
pub async fn sandbox_get_state(
    sandbox: State<'_, SandboxManager>,
) -> Result<SandboxState, String> {
    Ok(sandbox.state().await)
}

#[tauri::command]
pub async fn sandbox_start(
    app: AppHandle,
    sandbox: State<'_, SandboxManager>,
) -> Result<(), String> {
    sandbox.auto_start(&app).await
}

#[tauri::command]
pub async fn sandbox_stop(
    app: AppHandle,
    sandbox: State<'_, SandboxManager>,
) -> Result<(), String> {
    sandbox.stop(&app).await
}

#[tauri::command]
pub async fn sandbox_restart(
    app: AppHandle,
    sandbox: State<'_, SandboxManager>,
) -> Result<(), String> {
    sandbox.restart(&app).await
}

#[tauri::command]
pub async fn sandbox_exec(
    command: String,
    cwd: String,
    sandbox: State<'_, SandboxManager>,
) -> Result<ExecResult, String> {
    let cwd = if cwd.is_empty() {
        "/workspace".to_string()
    } else {
        cwd
    };
    sandbox.exec(&command, &cwd, 30).await
}

#[tauri::command]
pub async fn sandbox_mount_workspace(
    path: String,
    app: AppHandle,
    sandbox: State<'_, SandboxManager>,
) -> Result<(), String> {
    sandbox.mount_workspace(&path, &app).await
}

#[tauri::command]
pub async fn sandbox_clean_workspace(
    sandbox: State<'_, SandboxManager>,
) -> Result<(), String> {
    sandbox.clean_workspace().await
}

#[tauri::command]
pub async fn sandbox_set_enabled(
    enabled: bool,
    app: AppHandle,
    sandbox: State<'_, SandboxManager>,
) -> Result<(), String> {
    sandbox.set_enabled(enabled, &app).await
}
